src/textual/widgets/_directory_tree.py

```diff
@@ -229,37 +238,115 @@ def filter_paths(self, paths: Iterable[Path]) -> Iterable[Path]:
         """
         return paths
 
-    def _load_directory(self, node: TreeNode[DirEntry]) -> None:
-        """Load the directory contents for a given node.
+    @staticmethod
+    def _safe_is_dir(path: Path) -> bool:
+        """Safely check if a path is a directory.
 
         Args:
-            node: The node to load the directory contents for.
+            path: The path to check.
+
+        Returns:
+            `True` if the path is for a directory, `False` if not.
         """
-        assert node.data is not None
-        node.data.loaded = True
-        directory = sorted(
-            self.filter_paths(node.data.path.iterdir()),
-            key=lambda path: (not path.is_dir(), path.name.lower()),
-        )
-        for path in directory:
+        try:
+            return path.is_dir()
+        except PermissionError:
+            # We may or may not have been looking at a directory, but we
+            # don't have the rights or permissions to even know that. Best
+            # we can do, short of letting the error blow up, is assume it's
+            # not a directory. A possible improvement in here could be to
+            # have a third state which is "unknown", and reflect that in the
+            # tree.
+            return False
+
+    def _populate_node(self, node: TreeNode[DirEntry], content: Iterable[Path]) -> None:
+        """Populate the given tree node with the given directory content.
+
+        Args:
+            node: The Tree node to populate.
+            content: The collection of `Path` objects to populate the node with.
+        """
+        for path in content:
             node.add(
                 path.name,
                 data=DirEntry(path),
-                allow_expand=path.is_dir(),
+                allow_expand=self._safe_is_dir(path),
             )
         node.expand()
 
-    def _on_mount(self, _: Mount) -> None:
-        self._load_directory(self.root)
+    def _directory_content(self, location: Path, worker: Worker) -> Iterator[Path]:
+        """Load the content of a given directory.
+
+        Args:
+            location: The location to load from.
+            worker: The worker that the loading is taking place in.
+
+        Yields:
+            Path: A entry within the location.
+        """
+        try:
+            for entry in location.iterdir():
+                if worker.is_cancelled:
+                    break
+                yield entry
+        except PermissionError:
+            pass
+
+    @work
+    def _load_directory(self, node: TreeNode[DirEntry]) -> list[Path]:
+        """Load the directory contents for a given node.
+
+        Args:
+            node: The node to load the directory contents for.
+
+        Returns:
+            The list of entries within the directory associated with the node.
+        """
+        assert node.data is not None
+        node.data.loaded = True
```

Comment:
This may be the wrong place to set the `loaded` flag. It looks like there is theoretical window where you may end up loading the same directory more than once. 

I think you can avoid that by checking and setting the loaded flag outside of the worker.

Reply:
Aye, good cool, that can move up.